src/io/reader_cache.cc

```diff
@@ -0,0 +1,69 @@
+/*!
+ *
+﻿ * Copyright (C) 2015 Technical University of Liberec.  All rights reserved.
+ * 
+ * This program is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License version 3 as published by the
+ * Free Software Foundation. (http://www.gnu.org/licenses/gpl-3.0.en.html)
+ * 
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+ *
+ * 
+ * @file    reader_cache.cc
+ * @brief   
+ */
+
+#include "io/reader_cache.hh"
+#include "io/msh_gmshreader.h"
+#include "io/msh_vtkreader.hh"
+#include "io/msh_pvdreader.hh"
+#include "input/accessors.hh"
+
+
+/***********************************************************************************************
+ * Implementation of ReaderCache
+ */
+
+ReaderCache * ReaderCache::instance() {
+	static ReaderCache *instance = new ReaderCache;
+	return instance;
+}
+
+std::shared_ptr<BaseMeshReader> ReaderCache::get_reader(const FilePath &file_path) {
+	return ReaderCache::get_reader_data(file_path)->second.reader_;
+}
+
+std::shared_ptr<Mesh> ReaderCache::get_mesh(const FilePath &file_path) {
+	auto it = ReaderCache::get_reader_data(file_path);
+	// Create and fill mesh if doesn't exist
+	if ( (*it).second.mesh_ == nullptr ) {
+		(*it).second.mesh_ = std::make_shared<Mesh>( Input::Record() );
+		(*it).second.reader_->read_raw_mesh( (*it).second.mesh_.get() );
+		(*it).second.reader_->check_compatible_mesh( *((*it).second.mesh_) );
```

Comment:
- ? need to call setup_topology after read_raw_mesh, rather NO
- make node and element caches membeers of the VTU reader so that read_nodes and red_elements can reuse already loaded nodes and elements
- use read_nodes and read_elements in  check_compatible_mesh